### PR TITLE
Fix systemd time-wait-sync inotify bugs

### LINKED ISSUE #896: systemd-time-wait-sync.service gets stuck sometimes
I'm not sure if this is a systemd bug or kernel issue, but on RPI4 after a reboot I often notice that `systemd-time-wait-sync.service` has become stuck:

```
# systemctl list-jobs
JOB UNIT                                 TYPE  STATE
 69 e2scrub_all.timer                    start waiting
 38 time-sync.target                     start waiting
 67 timers.target                        start waiting
 79 systemd-update-utmp-runlevel.service start waiting
  1 multi-user.target                    start waiting
 37 systemd-time-wait-sync.service       start running
```

```
# systemctl status systemd-time-wait-sync.service
Warning: The unit file, source configuration file or drop-ins of systemd-time-wait-sync.service changed on disk. Run 'sys>
● systemd-time-wait-sync.service - Wait Until Kernel Time Synchronized
     Loaded: loaded (/usr/lib/systemd/system/systemd-time-wait-sync.service; enabled; vendor preset: enabled)
    Drop-In: /etc/systemd/system/systemd-time-wait-sync.service.d
             └─hassos.conf
     Active: activating (start) since Fri 2020-02-07 15:50:52 UTC; 8 months 4 days ago
       Docs: man:systemd-time-wait-sync.service(8)
   Main PID: 117 (systemd-time-wa)
      Tasks: 1 (limit: 2148)
     Memory: 440.0K
     CGroup: /system.slice/systemd-time-wait-sync.service
             └─117 /usr/lib/systemd/systemd-time-wait-sync

Feb 07 15:50:53 server systemd-time-wait-sync[117]: adjtime state 5 status 40 time Fri 2020-02-07 15:50:52.918974 UTC
Oct 12 19:37:06 server systemd-time-wait-sync[117]: adjtime state 5 status 40 time Mon 2020-10-12 19:37:06.002184 UTC
Oct 12 19:38:23 server systemd-time-wait-sync[117]: adjtime state 5 status 40 time Mon 2020-10-12 19:38:23.754433 UTC
Warning: Journal has been rotated since unit was started. Log output is incomplete or unavailable.
```

This happens even though the `systemd-timesyncd.service` is activated, and even though `/run/systemd/timesync/synchronized` exists:

```
# systemctl status systemd-timesyncd
Warning: The unit file, source configuration file or drop-ins of systemd-timesyncd.service changed on disk. Run 'systemct>
● systemd-timesyncd.service - Network Time Synchronization
     Loaded: loaded (/usr/lib/systemd/system/systemd-timesyncd.service; enabled; vendor preset: enabled)
    Drop-In: /etc/systemd/system/systemd-timesyncd.service.d
             └─hassos.conf, ro.conf
     Active: active (running) since Mon 2020-10-12 19:37:06 UTC; 26min ago
       Docs: man:systemd-timesyncd.service(8)
   Main PID: 252 (systemd-timesyn)
     Status: "Initial synchronization to time server 162.159.200.123:123 (time.cloudflare.com)."
      Tasks: 2 (limit: 2148)
     Memory: 1.2M
     CGroup: /system.slice/systemd-timesyncd.service
             └─252 /usr/lib/systemd/systemd-timesyncd

Feb 07 15:50:55 server systemd[1]: Starting Network Time Synchronization...
Feb 07 15:50:56 server systemd-timesyncd[252]: System clock time unset or jumped backwards, restoring from recorded times>
Oct 12 19:37:06 server systemd[1]: Started Network Time Synchronization.
Oct 12 19:37:06 server systemd-timesyncd[252]: Network configuration changed, trying to establish connection.
Oct 12 19:37:06 server systemd-timesyncd[252]: Network configuration changed, trying to establish connection.
Oct 12 19:38:23 server systemd-timesyncd[252]: Initial synchronization to time server 162.159.200.123:123 (time.cloudflar>
```

```
# ls -alh /run/systemd/timesync/synchronized
-rw-r--r--    1 systemd- systemd-       0 Oct 12 19:54 /run/systemd/timesync/synchronized
```

Further we can see that `systemd-time-wait-sync` is stuck waiting for events:

```
# lsof -nPp 117
COMMAND   PID USER   FD      TYPE             DEVICE SIZE/OFF  NODE NAME
systemd-t 117 root  cwd       DIR                8,3      238  7333 /
systemd-t 117 root  rtd       DIR                8,3      238  7333 /
systemd-t 117 root  txt       REG                8,3    14104  3808 /usr/lib/systemd/systemd-time-wait-sync
systemd-t 117 root  mem       REG                8,3    92368  1512 /usr/lib/libz.so.1.2.11
systemd-t 117 root  mem       REG                8,3   129824  1328 /usr/lib/libgpg-error.so.0.28.0
systemd-t 117 root  mem       REG                8,3   155792  1459 /usr/lib/libpthread-2.30.so
systemd-t 117 root  mem       REG                8,3   628680  1373 /usr/lib/libm-2.30.so
systemd-t 117 root  mem       REG                8,3    31352  1467 /usr/lib/librt-2.30.so
systemd-t 117 root  mem       REG                8,3   374464  1383 /usr/lib/libmount.so.1.1.0
systemd-t 117 root  mem       REG                8,3    88696  1369 /usr/lib/libkmod.so.2.3.4
systemd-t 117 root  mem       REG                8,3    31224  1340 /usr/lib/libip4tc.so.2.0.0
systemd-t 117 root  mem       REG                8,3   773704  1301 /usr/lib/libgcrypt.so.20.2.5
systemd-t 117 root  mem       REG                8,3    22856  1236 /usr/lib/libcap.so.2.27
systemd-t 117 root  mem       REG                8,3   324232  1227 /usr/lib/libblkid.so.1.1.0
systemd-t 117 root  mem       REG                8,3  1413888  1232 /usr/lib/libc-2.30.so
systemd-t 117 root  mem       REG                8,3  2220416  3496 /usr/lib/systemd/libsystemd-shared-244.so
systemd-t 117 root  mem       REG                8,3   159440  1199 /usr/lib/ld-2.30.so
systemd-t 117 root    0r      CHR                1,3      0t0  5418 /dev/null
systemd-t 117 root    1u     unix 0x0000000057e493d7      0t0 10285 type=STREAM
systemd-t 117 root    2u     unix 0x0000000057e493d7      0t0 10285 type=STREAM
systemd-t 117 root    3u  a_inode               0,12        0  5352 [eventpoll]
systemd-t 117 root    4u  a_inode               0,12        0  5352 [signalfd]
systemd-t 117 root    5r  a_inode               0,12        0  5352 inotify
systemd-t 117 root    6u  a_inode               0,12        0  5352 [timerfd]

# strace -ttTp 117
strace: Process 117 attached
20:08:13.919924 epoll_pwait(3,
```

Internally, `systemd-time-sync-wait` will either wait for a clock event from the kernel (which is known to be unreliable), or for an inotify event for `/run/systemd/timesync/synchronized`. However it seems the latter is not working, even if I `touch` the file again or delete and re-create it.

```
# strace -ttTp 117 &
# touch /run/systemd/timesync/synchronized
strace: Process 117 attached
20:11:47.916173 epoll_pwait(3, 
```

Interestingly, if I delete and re-create the the directory and then the file it gets an event but still doesn't activate:

```
# rm /run/systemd/timesync/synchronized ; rmdir /run/systemd/timesync ; mkdir /run/systemd/timesync ; touch /run/systemd/timesync/synchronized
[{EPOLLIN, {u32=907074160, u64=907074160}}], 4, -1, NULL, 8) = 1 <40.990304>
20:13:36.995021 clock_gettime(CLOCK_BOOTTIME, {tv_sec=2149, tv_nsec=284383941}) = 0 <0.000026>
20:13:36.995165 read(5, "\1\0\0\0\0\1\0@\0\0\0\0\20\0\0\0timesync\0\0\0\0\0\0\0\0", 272) = 32 <0.000028>
20:13:36.995273 epoll_pwait(3,
```

If we look at the inotify state of the daemon, it looks like only one watcher is setup:

```
# cat /proc/117/fdinfo/5
pos:	0
flags:	02004000
mnt_id:	12
inotify wd:1 ino:2014 sdev:15 mask:100 ignored_mask:0 fhandle-bytes:c fhandle-type:1 f_handle:824d969c1420000000000000
```

Looking at `mnt_id: 12` in mountinfo shows it is attached to the `/run` filesystem:

```
# cat /proc/117/mountinfo  | grep :12
26 19 0:21 / /run rw,nosuid,nodev shared:12 - tmpfs tmpfs rw,mode=755
40 19 0:21 /machine-id /etc/machine-id ro shared:12 - tmpfs tmpfs rw,mode=755
```

On the systemd project there are some references to similar problems. So it is possible this is a systemd bug, but the check for `/run/systemd/timesync/synchronized` was specifically added to combat this and I think maybe there is some bad interaction with inotify or the kernel version which is to blame.

https://github.com/systemd/systemd/issues/8683
https://github.com/systemd/systemd/issues/15947

### PR DESCRIPTION
**Pull request by tmm1**
fixes #896 

cc @agners

### CODE REVIEW

File: buildroot/package/systemd/systemd.mk [hunk 1]
```diff
@@ -41,7 +41,8 @@ SYSTEMD_CONF_OPTS += \
 	-Dumount-path=/usr/bin/umount \
 	-Dnobody-group=nogroup \
 	-Didn=true \
-	-Dnss-systemd=true
+	-Dnss-systemd=true \
+	-Dapparmor=false
```
Comment (tmm1) [inline buildroot/package/systemd/systemd.mk:45]: I'm not sure what the policy is regarding changes to the vendored buildroot.

Looks like upstream has already switched to systemd v246.5, and also has new flags to enable/disable apparmor:

https://github.com/buildroot/buildroot/blob/master/package/systemd/systemd.mk#L56-L61

Comment (agners) [inline buildroot/package/systemd/systemd.mk:45]: This change also needs to be a patch in buildroot-patches/ so we can update buildroot with scripts.

And ideally, also sent upstream to buildroot :wink:

Comment (agners) [reply buildroot/package/systemd/systemd.mk:45]: I see, in that case just add a patch to `buildroot-patches/` as well, kinda as a marker to check when we bump buildroot.

Comment (tmm1) [reply buildroot/package/systemd/systemd.mk]: Okay I will just remove this for now.

**Review by agners — APPROVED** [review]
Ok, adding patch in `buildroot-external/` is even nicer I agree. LGTM.

**Review by pvizeli — APPROVED** [review]

File: buildroot/package/systemd/systemd.mk (hunk 1 above)
Comment (pvizeli) [reply buildroot/package/systemd/systemd.mk]: Yeah buildroot upstream support now apparmor

**Pull request merged**